### ISSUE TITLE
caret wrapper failing with numerical target.

### ISSUE THREAD

**Issue by hadjipantelis** [catboost/catboost#35]
Hello,

Thank you very much for making `catboost` available and for all your work on this. Unfortunately the `catboost.caret` wrapper seems to fail even in very simple examples when using numerical responses. For example:

```R
library(caret)
require(catboost)

set.seed(12345)

target_ <- sample(c(1, 2, 3), size = 1000, replace = TRUE)

data <- data.frame(f_numeric = target_ + rnorm(length(target_), mean = 0, sd = 1),
                   f_logical = (target_ + rnorm(length(target_), mean = 0, sd = 1)) > 0,
                   f_factor = as.factor(round(10 * (target_ + rnorm(length(target_), mean = 0, sd = 1)))),
                   f_character = as.character(round(10 * (target_ + rnorm(length(target_), mean = 0, sd = 1)))))

data$f_logical = as.factor(data$f_logical)
data$f_character = as.factor(data$f_character)

data$target <- target_ # as.factor(make.names(target_))

fit_control <- trainControl(method = "cv",    number = 4,   classProbs = TRUE)

grid <- expand.grid(depth = c(4, 6), learning_rate = c(0.01, 0.1, 0.2), iterations = 10)

report <- train(target ~ f_numeric + f_logical + f_factor,    data = data,   
                       method = catboost.caret, verbose = FALSE,   preProc = NULL, 
                       tuneGrid = grid,   trControl = fit_control)
Something is wrong; all the RMSE metric values are missing:
      RMSE        Rsquared        MAE     
 Min.   : NA   Min.   : NA   Min.   : NA  
 1st Qu.: NA   1st Qu.: NA   1st Qu.: NA  
 Median : NA   Median : NA   Median : NA  
 Mean   :NaN   Mean   :NaN   Mean   :NaN  
 3rd Qu.: NA   3rd Qu.: NA   3rd Qu.: NA  
 Max.   : NA   Max.   : NA   Max.   : NA  
 NA's   :6     NA's   :6     NA's   :6     
```

In addition, would it be possible to add the `l2_leaf_reg` and `rsm` variable as being tunable? These two seem as the two most obvious parameters to control over-fitting. 
Thanks again for all your work on `catboost`!

**Comment by grayskripko** [comment]
Probably, it is an incorrect example. I got another error:
```
Error: The tuning parameter grid should have columns depth, learning_rate, iterations, l2_leaf_reg, rsm, border_count
In addition: Warning message:
In train.default(x, y, weights = w, ...) :
  cannnot compute class probabilities for regression
```

**Comment by grayskripko** [comment]
the problem is in caret:::train.formula transforms X to ``` x <- model.matrix(Terms, m, contrasts) ``` and the beginning of catboost.from_data_frame() contains 
```
if (!is.data.frame(data)) 
    stop("Unsupported data type, expecting data.frame, got: ", 
      class(data))
```

**Comment by grayskripko** [comment]
As a workaround, you should avoid using formula in caret::train. Also, pay attention to target variable class. 
Probably, you should uncomment ``` # as.factor(make.names(target_)) ``` for classification or remove classProbs=T for regression

**Comment by hadjipantelis** [comment]
@grayskripko Thank you very much for looking into this. 

The example was just the original `caret.R` in `demo` with numeric instead of factor response, it has been updated since. You got a different error now because @exprmntr updated `catboost.caret` (commit bcf5caf - thank you very much about that @exprmntr that was great response time! :+1: ).
Yes, obviously I would cast the `target` as factor for classification; sorry, I should have been explicit that I focused on a regression test case. (I just left `classProbs` in by accident, as you saw these are irrelevant warnings.)

Cool, thanks for the proposed work-around. It makes sense as a step towards the right direction. I will check the new regression code as soon as I can. 
BTW, the current example for classification while computing it seems a bit unstable. For example if someone `set.seed(34)` right before `report <- train(...` the run will fail; this seems particular to the folds used. ie.:

```R
set.seed(34)
report <- train(x, as.factor(make.names(y)),
                method = catboost.caret,
                verbose = TRUE, preProc = NULL,
                tuneGrid = grid, trControl = fit_control)

 Error in { : task 1 failed - "NA and NULL values are not supported: Age" In addition: Warning messages:
1: In eval(xpr, envir = envir) :
  predictions failed for Fold3: depth=4, learning_rate=0.1, iterations=100, l2_leaf_reg=0.001, rsm=0.95, border_count=64 Error in catboost.from_data_frame(newdata) : 
  NA and NULL values are not supported: Age
....
```
As a side-note: Is `train` supposed to be reproducible? For example even if I `set.seed` right before `train` in the classification example with a seed that works, say `set.seed(4)`, if I repeat it I do not get the same metrics, `varImp`, etc. Is that because of multi-threading being lock-free when parallelising? 
Once again, thanks for checking this issue out and thanks @exprmntr for the feature additions!

For the record my `sessionInfo()` is:

```R 
> sessionInfo()
R version 3.4.0 (2017-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 8 (jessie)

Matrix products: default
BLAS: /usr/lib/openblas-base/libblas.so.3
LAPACK: /usr/lib/libopenblasp-r0.2.12.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] titanic_0.1.0    catboost_0.1.1.5 jsonlite_1.5     caret_6.0-77     ggplot2_2.2.1    lattice_0.20-35 

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.12       compiler_3.4.0     nloptr_1.0.4       plyr_1.8.4         class_7.3-14       iterators_1.0.8   
 [7] tools_3.4.0        testthat_1.0.2     digest_0.6.12      lme4_1.1-13        memoise_1.1.0      tibble_1.3.3      
[13] gtable_0.2.0       nlme_3.1-131       mgcv_1.8-17        rlang_0.1.1        Matrix_1.2-10      foreach_1.4.3     
[19] commonmark_1.2     parallel_3.4.0     SparseM_1.77       e1071_1.6-8        withr_1.0.2        stringr_1.2.0     
[25] roxygen2_6.0.1     xml2_1.1.1         MatrixModels_0.4-1 devtools_1.13.2    stats4_3.4.0       grid_3.4.0        
[31] nnet_7.3-12        R6_2.2.2           minqa_1.2.4        reshape2_1.4.2     car_2.1-5          magrittr_1.5      
[37] ModelMetrics_1.1.0 codetools_0.2-15   scales_0.4.1       MASS_7.3-47        splines_3.4.0      pbkrtest_0.4-7    
[43] colorspace_1.3-2   quantreg_5.33      stringi_1.1.5      lazyeval_0.2.0     munsell_0.4.3      crayon_1.3.2      
```

**Comment by hadjipantelis** [comment]
@a4tunado Thank you.

**Comment by jamesmbull** [comment]
Sorry to comment on a closed item but I have updated everything from the git repo and I'm still getting:

`Unsupported data type, expecting data.frame, got: matrix`

... In a regression model. I found the post from @grayskripko searching for info about this error. Does the post by @hadjipantelis mean that #bcf5caf introduced a new error? 

`> sessionInfo()
R version 3.4.1 (2017-06-30)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Arch Linux

Matrix products: default
BLAS: /usr/lib/libblas.so.3.7.0
LAPACK: /usr/lib/liblapack.so.3.7.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Metrics_0.1.2     caret_6.0-76      ggplot2_2.2.1     lattice_0.20-35  
[5] data.table_1.10.4 catboost_0.1.1.5 

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.12       magrittr_1.5       splines_3.4.1      MASS_7.3-47       
 [5] munsell_0.4.3      colorspace_1.3-2   rlang_0.1.1        foreach_1.4.3     
 [9] minqa_1.2.4        stringr_1.2.0      car_2.1-5          plyr_1.8.4        
[13] tools_3.4.1        parallel_3.4.1     pbkrtest_0.4-7     nnet_7.3-12       
[17] grid_3.4.1         gtable_0.2.0       nlme_3.1-131       mgcv_1.8-17       
[21] quantreg_5.33      MatrixModels_0.4-1 iterators_1.0.8    lme4_1.1-13       
[25] lazyeval_0.2.0     tibble_1.3.3       Matrix_1.2-10      nloptr_1.0.4      
[29] reshape2_1.4.2     ModelMetrics_1.1.0 codetools_0.2-15   stringi_1.1.5     
[33] compiler_3.4.1     scales_0.4.1       stats4_3.4.1       SparseM_1.77 `

My code is below. Thanks for any additional insight and thanks for the work on catboost and caret integration!

```
require(catboost)
require(data.table)
require(caret)
require(Metrics) # for MAE

### LOAD AND TWEAK DATA ###

# Don't use scientific notation in output
options(scipen = 999)

#### LOAD AND BASIC SET-UP ####

pr <- fread('properties_2016.csv')

t <- fread('train_2016_v2.csv')

# Drop dupes
t <- t[!duplicated(t$parcelid), ]

t[,transactiondate := NULL]

# List of logical vectors
logilist <- c(
            "architecturalstyletypeid",
            "hashottuborspa",
            "pooltypeid10",
            "pooltypeid2",
            "typeconstructiontypeid",
            "fireplaceflag"
            )

# Convert logicals to factors
pr[, logilist] <- lapply(pr[, logilist, with=FALSE], factor)

# Convert char to factor
pr$propertycountylandusecode <- as.factor(pr$propertycountylandusecode)
pr$propertyzoningdesc <- as.factor(pr$propertyzoningdesc)
pr$taxdelinquencyflag <- as.factor(pr$taxdelinquencyflag)

# Drop
pr[,censustractandblock := NULL]

# Considering imputing NAs or just want to speed things up? 
# Drop all vectors with >= x% NA
vectordrop <- pr[, lapply( pr, function(x) sum(is.na(x)) / length(x) ) >= .30 ]

pr[, names(which(vectordrop == TRUE)) := NULL]

imp_values <- preProcess(pr,
                        method = c("medianImpute")
                        )

p <- predict(imp_values, pr)

#### BUILD MODEL ####

# Create training set
train <- merge(p, t, by="parcelid", all.y=TRUE)

# We need all of the RAM we can get
rm(t)

p[, logerror := 0]

set.seed(28)

# Enable caret to use MAE as eval metric
maeSummary <- function (train,
                        lev = NULL,
                        model = NULL) {
                            out <- mae(train$obs, train$pred)
                            names(out) <- "MAE"
                            out
                        }

control <- trainControl(method = "cv",
                              number = 3,
                              verboseIter=TRUE,
                              summaryFunction = maeSummary
                              )

grid <- expand.grid(depth = c(3, 4),
                    learning_rate = 0.1,
                    iterations = 10,
                    l2_leaf_reg = 1e-3,
                    rsm = 0.95,
                    border_count = 32)

cb <- train(logerror ~ .,
            data=train,
            preProcess=NULL,
            method=catboost.caret,
            metric = "MAE",
            maximize = FALSE,
            tuneGrid = grid,
            trControl = control
            )

print(cb)

#importance <- varImp(cb, scale = FALSE)
#print(importance)

cb_prediction <- predict(cb, p)

cb_submit <- data.frame(ParcelId=p$parcelid,
                        X201610=round(cb_prediction,4),
                        X201611=round(cb_prediction,4),
                        X201612=round(cb_prediction,4),
                        X201710=0,
                        X201711=0,
                        X201712=0
                        )

write.csv(cb_submit, file = "cb-prediction.csv", row.names = FALSE)
```

**Comment by grayskripko** [comment]
@jamesmbull it looks like the current version still contains this bug https://github.com/catboost/catboost/blob/347d0f5341915f4738aee208d03d7bd3f3fc6650/catboost/R-package/R/catboost.R#L100
I did not notice any changes according to the problem. Just don't use formula in train(). By the way, I heard it is a little bit faster to use train(..., x =, y =...)
One more thought. It looks alarming when you name your variable 'train' and use caret::train in the same time. May produce bad bugs

**Comment by jamesmbull** [comment]
@grayskripko, thank you for the response and I'll take your advice! Specifically, what do you mean by using formula in train?

**Comment by hadjipantelis** [comment]
@jamesmbull By "use formula to train" one would mean the interface `train(y ~ x1 + x2 + xwhatever, data = myData, ...)`. It appears that constructing the de matrix can be finicky so actually passing the matrix `X` along with `y` is cleaner and potentially faster. (ie. `train( y=y, x=X, ...)`.  bcf5caf among other things changed the `catboost.caret` to use more tuning parameters but did not explicitly changed something about the error mentioned.

And yeah, as @grayskripko mentioned using `train` as a variable name while it already exists as a `method` can be [bad practice](https://stackoverflow.com/questions/6135868), it is not catastrophic but why risk it?
